### PR TITLE
feat(retrofit): Replace RetrofitError catch blocks and add spinnakerE…

### PR DESCRIPTION
**Pull request by DanielaS12**
This pull request registers SpinnakerRetrofitErrorHandler with each Retrofit.RestAdapter and replaces each RetrofitError catch block with a catch-block using SpinnakerServerException or the appropriate subclass. This is part of an effort to consume SpinnakerRetrofitErrorHandler in each Spinnaker microservice, as detailed in https://github.com/spinnaker/spinnaker/issues/5473

### CODE REVIEW

**Comment by dbyron-sf** [comment]
@DanielaS12 Looks like some tests are failing.

**Comment by dbyron-sf** [comment]
This would be easier to follow as separate commits -- one per retrofit interface.

File: clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java [hunk 1]
```diff
@@ -15,11 +15,7 @@
  */
 package com.netflix.spinnaker.clouddriver.aws.lifecycle;
 
-import com.amazonaws.auth.policy.Condition;
-import com.amazonaws.auth.policy.Policy;
-import com.amazonaws.auth.policy.Principal;
-import com.amazonaws.auth.policy.Resource;
-import com.amazonaws.auth.policy.Statement;
+import com.amazonaws.auth.policy.*;
```
Comment (dbyron-sf) [inline clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java:18]: please avoid star imports

File: clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java [hunk 2]
```diff
@@ -227,24 +218,31 @@ private boolean updateInstanceStatus(Eureka eureka, String app, String instanceI
...
-        } else {
-          log.error(
-              "Irrecoverable error while marking app out of service (app: {}, instance: {}, retry: {})",
-              app,
-              instanceId,
-              retry,
-              e);
-          break;
-        }
+      } catch (NotFoundException e) {
```
Comment (dbyron-sf) [inline clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java:221]: SpinnakerHttpException with a 404 response code is the equivalent of RetrofitError with a 404 response code -- not NotFoundException.

File: clouddriver-consul/src/main/groovy/com/netflix/spinnaker/clouddriver/consul/api/v1/ConsulAgent.groovy [hunk 3]
```diff
@@ -19,11 +19,6 @@ package com.netflix.spinnaker.clouddriver.consul.api.v1
 import com.netflix.spinnaker.clouddriver.consul.api.v1.services.AgentApi
 import com.netflix.spinnaker.clouddriver.consul.config.ConsulConfig
 import com.netflix.spinnaker.clouddriver.consul.config.ConsulProperties
-import com.squareup.okhttp.OkHttpClient
-import retrofit.RestAdapter
-import retrofit.client.OkClient
-
-import java.util.concurrent.TimeUnit
```
Comment (dbyron-sf) [inline clouddriver-consul/src/main/groovy/com/netflix/spinnaker/clouddriver/consul/api/v1/ConsulAgent.groovy:26]: If there are some unrelated refactors, please separate them into their own commits to make this easier to review.

File: clouddriver-consul/src/main/groovy/com/netflix/spinnaker/clouddriver/consul/config/ConsulConfig.groovy [hunk 4]
```diff
@@ -57,7 +52,7 @@ class ConsulConfig {
       try {
         def catalog = new ConsulCatalog(this)
         datacenters = catalog.api.datacenters()
-      } catch (RetrofitError e) {
+      } catch (SpinnakerNetworkException e) {
```
Comment (dbyron-sf) [inline clouddriver-consul/src/main/groovy/com/netflix/spinnaker/clouddriver/consul/config/ConsulConfig.groovy:55]:
```suggestion
      } catch (SpinnakerServerException e) {
```

File: clouddriver-consul/src/main/groovy/com/netflix/spinnaker/clouddriver/consul/provider/ConsulProviderUtils.groovy [hunk 5]
```diff
@@ -40,7 +40,7 @@ class ConsulProviderUtils {
         return new ConsulService(result)
       } ?: []
       running = true
-    } catch (RetrofitError e) {
+    } catch (SpinnakerHttpException e) {
```
Comment (dbyron-sf) [inline clouddriver-consul/src/main/groovy/com/netflix/spinnaker/clouddriver/consul/provider/ConsulProviderUtils.groovy:43]:
```suggestion
    } catch (SpinnakerServerException e) {
```

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/client/DockerRegistryClient.groovy [hunk 6]
```diff
@@ -444,10 +442,10 @@ class DockerRegistryClient {
   public void checkV2Availability() {
     try {
       doCheckV2Availability()
-    } catch (RetrofitError error) {
+    } catch (SpinnakerHttpException error) {
```
Comment (dbyron-sf) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/api/v2/client/DockerRegistryClient.groovy:445]: I think this needs to be SpinnakerServerException with some additional instanceof checks for SpinnakerHttpException to get the same behavior as before.

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/provider/agent/DockerRegistryImageCachingAgent.groovy [hunk 7]
```diff
@@ -96,10 +97,9 @@ class DockerRegistryImageCachingAgent implements CachingAgent, AccountAware, Age
       DockerRegistryTags tags = null
       try {
         tags = credentials.client.getTags(repository)
-      } catch (Exception e) {
-        if (e instanceof RetrofitError && e.response?.status == 404) {
-          log.warn("Could not load tags for ${repository} in ${credentials.client.address}, reason: ${e.message}")
-        } else {
+      } catch (NotFoundException e) {
```
Comment (dbyron-sf) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/provider/agent/DockerRegistryImageCachingAgent.groovy:100]:
```suggestion
      } catch (SpinnakerHttpException e) {
```
with logic to check for a 404 response code.

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/provider/agent/DockerRegistryImageCachingAgent.groovy [hunk 8]
```diff
@@ -147,18 +146,16 @@ class DockerRegistryImageCachingAgent implements CachingAgent, AccountAware, Age
...
           try {
             digest = credentials.client.getDigest(repository, tag)
-          } catch (Exception e) {
-            if (e instanceof RetrofitError && ((RetrofitError) e).response?.status == 404) {
-              // Indicates inconsistency in registry, or deletion between call for all tags and manifest retrieval.
-              // In either case, we need to trust that this tag no longer exists.
-              log.warn("Image manifest for $tagKey no longer available; tag will not be cached: $e.message")
-              return
-            } else {
+          } catch (NotFoundException e) {
```
Comment (dbyron-sf) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/provider/agent/DockerRegistryImageCachingAgent.groovy:149]:
```suggestion
          } catch (SpinnakerHttpException e) {
```
with logic to check for a 404 response code.

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy [hunk 9]
```diff
@@ -426,13 +427,11 @@ class DockerRegistryNamedAccountCredentials extends AbstractAccountCredentials<D
...
 
-      return new DockerRegistryCredentials(client, repositories, trackDigests,  inspectDigests, skip, sortTagsByDate)
-    } catch (RetrofitError e) {
-      if (e.response?.status == 404) {
-        throw new DockerRegistryConfigException("No repositories specified for ${name}, and the provided endpoint ${address} does not support /_catalog.")
-      } else {
-        throw e
-      }
+      return new DockerRegistryCredentials(client, repositories, trackDigests, inspectDigests, skip, sortTagsByDate)
+    } catch (NotFoundException ignored) {
```
Comment (dbyron-sf) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy:431]:
```suggestion
    } catch (SpinnakerHttpException ignored) {
```
with logic to check for a 404 response code

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy [hunk 10]
```diff
@@ -426,13 +427,11 @@ class DockerRegistryNamedAccountCredentials extends AbstractAccountCredentials<D
...
+      throw new DockerRegistryConfigException("No repositories specified for ${name}, and the provided endpoint ${address} does not support /_catalog.")
+    } catch (SpinnakerException e) {
```
Comment (dbyron-sf) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy:433]: No need for a catch block that only throws.

File: clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy [hunk 11]
```diff
@@ -152,17 +156,20 @@ abstract class AbstractEurekaSupport {
             throw new RetryableException("Non HTTP 200 response from discovery for instance ${instanceId}, will retry (attempt: $retryCount}).")
           }
         }
-      } catch (RetrofitError retrofitError) {
-        def alwaysSkippable = retrofitError.response?.status == 404
-        def willSkip = alwaysSkippable || !strict
-        def skippingOrNot = willSkip ? "skipping" : "not skipping"
+      } catch (NotFoundException ignored) {
```
Comment (dbyron-sf) [inline clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy:159]:
```suggestion
      } catch (SpinnakerHttpException ignored) {
```
with logic to check for a 404 response code.

File: clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy [hunk 12]
```diff
@@ -219,17 +226,17 @@ abstract class AbstractEurekaSupport {
 
         retryCount++
         sleep(getDiscoveryRetryMs());
-      } catch (RetrofitError re) {
+      } catch (SpinnakerException re) {
```
Comment (dbyron-sf) [inline clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy:229]:
```suggestion
      } catch (SpinnakerServerException re) {
```

File: clouddriver-consul/src/main/groovy/com/netflix/spinnaker/clouddriver/consul/api/v1/ConsulAgent.groovy (hunk 3 above)
Comment (DanielaS12) [reply clouddriver-consul/src/main/groovy/com/netflix/spinnaker/clouddriver/consul/api/v1/ConsulAgent.groovy:26]: they are not unrelated. I removed unused retrofit imports.

File: clouddriver-appengine/clouddriver-appengine.gradle [hunk 13]
```diff
@@ -22,6 +22,7 @@ dependencies {
   implementation "io.spinnaker.kork:kork-config"
   implementation "io.spinnaker.kork:kork-cloud-config-server"
   implementation "io.spinnaker.kork:kork-moniker"
+  implementation "io.spinnaker.kork:kork-retrofit"
   implementation "com.netflix.spectator:spectator-api"
   implementation "com.squareup.okhttp:okhttp"
   implementation "com.squareup.retrofit:retrofit"
```
Comment (dbyron-sf) [inline clouddriver-appengine/clouddriver-appengine.gradle:28]: Can this disappear now?

File: clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java [hunk 14]
```diff
@@ -42,21 +38,16 @@
...
 import java.io.IOException;
 import java.time.Duration;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.Date;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
```
Comment (dbyron-sf) [inline clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java:46]: more star imports to undo

File: clouddriver-appengine/clouddriver-appengine.gradle (hunk 13 above)
Comment (DanielaS12) [reply clouddriver-appengine/clouddriver-appengine.gradle:28]: no, I am using SpinnakerRetrofitErrorHandler in AppengineConfigurationProperties

Comment (dbyron-sf) [reply clouddriver-appengine/clouddriver-appengine.gradle:28]: Sure, but SpinnakerRetrofitErrorHandler comes from kork-retrofit.  I'm asking about com.squareup.retrofit:retrofit.  If there are no more references to classes from there (e.g. RetrofitError), it can disappear.

File: clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java [hunk 15]
```diff
@@ -227,13 +228,22 @@ private boolean updateInstanceStatus(Eureka eureka, String app, String instanceI
...
-      } catch (RetrofitError e) {
-        final String recoverableMessage =
-            "Failed marking app out of service (status: {}, app: {}, instance: {}, retry: {})";
-        if (HttpStatus.NOT_FOUND.value() == e.getResponse().getStatus()) {
-          log.warn(recoverableMessage, e.getResponse().getStatus(), app, instanceId, retry);
-        } else if (e.getKind() == RetrofitError.Kind.NETWORK) {
-          log.error(recoverableMessage, e.getResponse().getStatus(), app, instanceId, retry, e);
+      } catch (SpinnakerServerException e) {
+        if (e instanceof SpinnakerHttpException
+            && ((SpinnakerHttpException) e).getResponseCode() == 404) {
```
Comment (dbyron-sf) [inline clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java:233]: Let's try to minimize diffs
```suggestion
            && HttpStatus.NOT_FOUND.value() == ((SpinnakerHttpException) e).getResponseCode()) {
```

File: clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java [hunk 16]
```diff
@@ -227,13 +228,22 @@ private boolean updateInstanceStatus(Eureka eureka, String app, String instanceI
       try {
         eureka.updateInstanceStatus(app, instanceId, DiscoveryStatus.OUT_OF_SERVICE.getValue());
         return true;
-      } catch (RetrofitError e) {
-        final String recoverableMessage =
```
Comment (dbyron-sf) [inline clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java:231]: Leaving this in minimizes diffs, and reduces duplication, yes?

File: clouddriver-consul/src/main/groovy/com/netflix/spinnaker/clouddriver/consul/api/v1/ConsulAgent.groovy (hunk 3 above)
Comment (dbyron-sf) [reply clouddriver-consul/src/main/groovy/com/netflix/spinnaker/clouddriver/consul/api/v1/ConsulAgent.groovy:26]: Side note -- please don't resolve conversations that you didn't initiate.

To me these are unrelated.  As in, they don't relate the commit message.  And I'm sticking to my story that smaller commits are easier to review.

File: clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy [hunk 17]
```diff
@@ -152,8 +154,8 @@ abstract class AbstractEurekaSupport {
             throw new RetryableException("Non HTTP 200 response from discovery for instance ${instanceId}, will retry (attempt: $retryCount}).")
           }
         }
-      } catch (RetrofitError retrofitError) {
-        def alwaysSkippable = retrofitError.response?.status == 404
+      } catch (SpinnakerHttpException e) {
```
Comment (dbyron-sf) [inline clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy:157]: Did I already comment about this elsewhere?  Seems like this needs to be 
```suggestion
      } catch (SpinnakerServerException e) {
```
to retain the previous behavior.

File: clouddriver-appengine/clouddriver-appengine.gradle (hunk 13 above)
Comment (DanielaS12) [reply clouddriver-appengine/clouddriver-appengine.gradle:28]: sure

File: clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java (hunk 16 above)
Comment (DanielaS12) [reply clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java:231]: yes

Comment (DanielaS12) [reply clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/lifecycle/InstanceTerminationLifecycleWorker.java:231]: will add it back

File: clouddriver-consul/src/main/groovy/com/netflix/spinnaker/clouddriver/consul/api/v1/ConsulAgent.groovy (hunk 3 above)
Comment (DanielaS12) [reply clouddriver-consul/src/main/groovy/com/netflix/spinnaker/clouddriver/consul/api/v1/ConsulAgent.groovy]: I added back the imports

File: clouddriver-appengine/clouddriver-appengine.gradle [hunk 18]
```diff
@@ -22,9 +22,9 @@ dependencies {
   implementation "io.spinnaker.kork:kork-config"
   implementation "io.spinnaker.kork:kork-cloud-config-server"
   implementation "io.spinnaker.kork:kork-moniker"
+  implementation "io.spinnaker.kork:kork-retrofit"
   implementation "com.netflix.spectator:spectator-api"
   implementation "com.squareup.okhttp:okhttp"
-  implementation "com.squareup.retrofit:retrofit"
```
Comment (dbyron-sf) [inline clouddriver-appengine/clouddriver-appengine.gradle:27]: Sorry if I asked you to remove this.  It needs to stay for e.g. [RetrofitAdapter](https://github.com/spinnaker/clouddriver/blob/6240ce89ff1564787c9391804db3131d92035e12/clouddriver-appengine/src/main/java/com/netflix/spinnaker/clouddriver/appengine/config/AppengineConfigurationProperties.java#L30C2-L30C29).

File: clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy [hunk 19]
```diff
@@ -219,17 +221,19 @@ abstract class AbstractEurekaSupport {
 
         retryCount++
         sleep(getDiscoveryRetryMs());
-      } catch (RetrofitError re) {
+      } catch (SpinnakerServerException re) {
```
Comment (dbyron-sf) [inline clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy:224]: nit:
```suggestion
      } catch (SpinnakerServerException e) {
```

File: clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy [hunk 20]
```diff
@@ -219,17 +221,19 @@ abstract class AbstractEurekaSupport {
...
         if (retryCount >= (maxRetries - 1)) {
           throw re
         }
 
-        AbstractEurekaSupport.log.debug("[$phaseName] - Failed calling external service ${re.message}")
+        AbstractEurekaSupport.log.debug("[$phaseName] - Failed calling external service ${re.getMessage()}")
 
-        if (re.kind == RetrofitError.Kind.NETWORK || re.response.status == 404 || re.response.status == 406) {
+        if ( (re instanceof SpinnakerNetworkException == true)
```
Comment (dbyron-sf) [inline clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy:231]: can't we do without the `== true` here?

File: clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy (hunk 19 above)
Comment (DanielaS12) [reply clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy:224]: addressed

File: clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy (hunk 20 above)
Comment (DanielaS12) [reply clouddriver-eureka/src/main/groovy/com/netflix/spinnaker/clouddriver/eureka/deploy/ops/AbstractEurekaSupport.groovy:231]: yes, we can

**Review by dbyron-sf — APPROVED** [review]

**Comment by link108** [comment]
@mergifyio update
